DSharpPlus.Core/Rest/Ratelimiting/RatelimitBucket.cs

```diff
@@ -0,0 +1,142 @@
+        public int Remaining
+        {
+            get => Volatile.Read(ref _remaining);
+            set => Volatile.Write(ref _remaining, value);
+        }
+#pragma warning restore CS0420
+
+        private volatile int _remaining = DefaultRemaining;
+
+        // this can and will wildly change
+        public DateTimeOffset ResetTime { get; private set; } = discordEpoch;
+
+        // and this is what we use to keep track of buckets
+        public string Hash { get; } = UnlimitedRatelimitIdentifier;
+
+        #region Constructing
+
+        public RatelimitBucket(int limit, int remaining, DateTimeOffset resetTime, string? hash)
+        {
+            Limit = limit;
+            Remaining = remaining;
+            ResetTime = resetTime;
+            Hash = hash ?? UnlimitedRatelimitIdentifier;
+        }
+
+        public static bool TryExtractRatelimitBucket(HttpResponseHeaders headers,
+            
+            [NotNullWhen(true)]
+            out RatelimitBucket? bucket)
+        {
+            bucket = null;
+
+            try
+            {
+                // these three headers are critical for us. if they don't exist, assume no valid ratelimit is passed
+                // and abort the operation, indicating that no ratelimit was extracted successfully.
+
+                if (!headers.TryGetValues("X-RateLimit-Limit", out IEnumerable<string>? limitRaw)
+                || !headers.TryGetValues("X-RateLimit-Remaining", out IEnumerable<string>? remainingRaw)
+                || !headers.TryGetValues("X-RateLimit-Reset", out IEnumerable<string>? ratelimitResetRaw))
+                {
+                    return false;
+                }
+
+                // again, these headers are critical. if they don't hold valid values, abort as well.
+
+                if (!int.TryParse(limitRaw.SingleOrDefault(), out int limit)
+                    || !int.TryParse(remainingRaw.SingleOrDefault(), out int remaining)
+                    || !double.TryParse(ratelimitResetRaw.SingleOrDefault(), out double ratelimitReset))
+                {
+                    return false;
+                }
+
+                string? hash = headers.GetValues("X-RateLimit-Name").SingleOrDefault();
+                DateTimeOffset resetTime = DateTimeOffset.UnixEpoch + TimeSpan.FromSeconds(ratelimitReset);
+
+                bucket = new(limit, remaining, resetTime, hash);
+                return true;
+            }
+            catch
+            {
+                return false;
+            }
+        }
+
+        #endregion
+
+        public void ResetBucket(DateTimeOffset nextResetTime)
+        {
+            if(nextResetTime < ResetTime)
+            {
+                throw new ArgumentOutOfRangeException(nameof(nextResetTime), "The next ratelimit bucket expiration cannot be in the past.");
+            }
+
+            lock (__lock)
```

Comment:
Now that I think about it, `lock`s aren't great when you're working with async, which is probably why Remora uses `SemaphoreSlim`s. Consider?

Reply:
The entire `RatelimitBucket` tries to avoid any kind of async logic. It is called asynchronously, but in itself works synchronously, which afaik `lock`s are fine for. If anything, I'd rather use `[MethodImpl(MethodImplOptions.Synchronized)]`. Allow me to benchmark for a second

Reply:
`lock` - 41 IL instructions, two expensive calls into the thread monitor, slow.
`SemaphoreSlim` - 31 IL instructions, two expensive calls into the thread monitor, two more less expensive calls into the thread monitor, a little faster.
`MethodImpl` attribute - 23 IL instructions, one call into the thread monitor, relatively fast

Reply:
switched to `MethodImpl` attribute; resolved